drivers/mbox/mbox_nxp_fsl_mu.c

```diff
@@ -0,0 +1,202 @@
+/*
+ * Wrapper of the Freescale Message Unit driver into Zephyr's MBOX model.
+ */
+
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/mbox.h>
+#include <zephyr/sys/util_macro.h>
+#include <fsl_mu.h>
+
+#define LOG_LEVEL CONFIG_MBOX_LOG_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(nxp_fsl_mu);
+
+#define DT_DRV_COMPAT nxp_fsl_mu
+
+#define MU_MAX_CHANNELS 2
+
+#define MU_MAX_MBOX_PER_CHAN	1
+#define MU_MBOX_SIZE		4
+
+#define MU_NODE(n)             DT_NODELABEL(mu##n)
+#define MU_BASE(n)             ((MU_Type *)DT_REG_ADDR(MU_NODE(n)))
+
+struct nxp_fsl_mu_data {
+	mbox_callback_t cb[MU_MAX_CHANNELS];
+	void *user_data[MU_MAX_CHANNELS];
+};
+
+struct nxp_fsl_mu_config {
+	MU_Type *base;
+	void (*config_irq)(void);
+};
+
+static int nxp_fsl_mu_send(const struct device *dev, uint32_t channel,
+			    const struct mbox_msg *msg)
+{
+	const struct nxp_fsl_mu_config *cfg = dev->config;
+	uint32_t data;
+
+	if (channel >= MU_MAX_CHANNELS) {
+		return -EINVAL;
+	}
+
+	/* Signalling mode. */
+	if (msg == NULL) {
+		if (kStatus_Success != MU_TriggerInterrupts(
+			cfg->base, kMU_GenInt0InterruptTrigger)) {
+			/* Previous interrupt has not been processed. */
+			return -EBUSY;
+		}
+		return 0;
+	}
+
+	/* Data transfer mode. */
+	data = *(uint32_t *)msg->data;
+	if (msg->size > (MU_MBOX_SIZE * MU_MAX_MBOX_PER_CHAN)) {
+		return -EMSGSIZE;
+	}
+
+	for (uint32_t i = 0; i < msg->size; ++i) {
+		MU_SendMsg(cfg->base, i, data);
+	}
+	return 0;
+}
+
+static int nxp_fsl_mu_register_callback(const struct device *dev, uint32_t channel,
+					 mbox_callback_t cb, void *user_data)
+{
+	struct nxp_fsl_mu_data *data = dev->data;
+
+	if (channel >= MU_MAX_CHANNELS) {
+		return -EINVAL;
+	}
+
+	data->cb[channel] = cb;
+	data->user_data[channel] = user_data;
+
+	return 0;
+}
+
+static int nxp_fsl_mu_mtu_get(const struct device *dev)
+{
+	return (MU_MBOX_SIZE * MU_MAX_MBOX_PER_CHAN);
+}
+
+static uint32_t nxp_fsl_mu_max_channels_get(const struct device *dev)
+{
+	return MU_MAX_CHANNELS;
+}
+
+static int nxp_fsl_mu_set_enabled(const struct device *dev, uint32_t channel,
+				   bool enable)
+{
+	struct nxp_fsl_mu_data *data = dev->data;
+	const struct nxp_fsl_mu_config *cfg = dev->config;
+
+	if (channel >= MU_MAX_CHANNELS) {
+		return -EINVAL;
+	}
+
+	if (enable && (data->cb[channel] == NULL)) {
+		LOG_WRN("Enabling channel without a registered callback\n");
+	}
+
+	if (enable) {
+		MU_EnableInterrupts(cfg->base,
+			(kMU_GenInt0InterruptEnable | kMU_Rx0FullInterruptEnable));
```

Comment:
Why do you enable/disable only general purpose interrupt 0 and RX0?

Reply:
General purpose interrupt 0 and RX0 are the minimum we need to support running Zephyr IPC service over this hardware. Doesn't have to be index 0, could be any other general purpose interrupt and RX. Enabling the other RXs can let this MBOX device send more data at once, I am ok with either just RX0 or more. Enabling other general interrupts is not helpful I think, since the [IPC service](https://docs.zephyrproject.org/latest/services/ipc/ipc_service/ipc_service.html) only needs one mechanism to notify the other side.

Reply:
I understand is working. This is great! :)
But, there is no consistency here.
You defined the MU_MBOX_SIZE and MU_MAX_MBOX_PER_CHAN ([here](https://github.com/zephyrproject-rtos/zephyr/pull/61910/files#diff-d617dd790b24bcbd937ef8d7508a31367c45869e809006a2f89f2fcf6d3f181dR18)) and based on these you should enable the interrupts.
So, in this case, you should enable all interrupts.

See an example here  - the legacy [IPM_IMX](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/ipm/ipm_imx.c#L253) which uses the same MU.